x-pack/plugin/core/src/main/java/org/elasticsearch/license/License.java

```diff
@@ -546,8 +558,9 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
         if (expiryDate != LicenseService.BASIC_SELF_GENERATED_LICENSE_EXPIRATION_MILLIS) {
             builder.timeField(Fields.EXPIRY_DATE_IN_MILLIS, Fields.EXPIRY_DATE, expiryDate);
         }
-
-        if (version >= VERSION_ENTERPRISE) {
+        if (hideEnterprise && builder.getRestApiVersion().matches(onOrAfter(RestApiVersion.V_7))) {
+            builder.field(Fields.MAX_NODES,maxNodes == -1  ? maxResourceUnits : maxNodes);
+        }else if (version >= VERSION_ENTERPRISE ) {
```

Comment:
Maybe I don't fully understand the intended behaviour for `hideEnterprise`. In [GetLicense 7.x](https://github.com/elastic/elasticsearch/blob/bad410f50982694eeab38f5610418014d3d9ac8e/x-pack/plugin/core/src/main/java/org/elasticsearch/license/License.java#L560-L567), it seems that it is only honored if `version < VERSION_ENTERPRISE`.  But this is already inconsistent with [RestXPackInfo 7.x](https://github.com/elastic/elasticsearch/blob/bad410f50982694eeab38f5610418014d3d9ac8e/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rest/action/RestXPackInfoAction.java#L45-L46) which always hide enterprise if requested.

Anyway, if we just want to keep whatever 7.x have as is and be compatible, I think we can just copy the code from 7.x, i.e.:
```java
if (version >= VERSION_ENTERPRISE) {
    builder.field(Fields.MAX_NODES, maxNodes == -1 ? null : maxNodes);
    builder.field(Fields.MAX_RESOURCE_UNITS, maxResourceUnits == -1 ? null : maxResourceUnits);
} else if (hideEnterprise && maxNodes == -1) {
    builder.field(Fields.MAX_NODES, maxResourceUnits);
} else {
    builder.field(Fields.MAX_NODES, maxNodes);
}
```
The value of `hideEnterprise` is already a result of checking rest compat, so I don't think we need check it again here.

Reply:
We discussed lively and realised that 7.x sets the [license version is set to before enterprise](https://github.com/elastic/elasticsearch/blob/abe59c3b019cb941904724e11787c68720e86120/x-pack/plugin/core/src/main/java/org/elasticsearch/license/RestGetLicenseAction.java#L57) when `accept_enterprise=false`. So it does always hide it similar to RestGetXPackInfo. 
We also decided that the V7 compat should just behave the same way as 7.x (by mostly adding similar code from 7.x).